current_proto.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"os"
+
+	"github.com/emicklei/proto"
+)
+
+type ProtoMessageMap map[string]*ProtoMessage
+
+func BuildCurrentProtoMap(filename string) (ProtoMessageMap, error) {
+	if filename == "" {
+		return ProtoMessageMap{}, nil
+	}
+	reader, _ := os.Open(filename)
+	defer reader.Close()
+
+	parser := proto.NewParser(reader)
+	definition, _ := parser.Parse()
+
+	p := make(ProtoMessageMap)
+	proto.Walk(definition, proto.WithMessage(p.HandleMessage()))
+
+	return p, nil
+}
+
+func (p ProtoMessageMap) HandleMessage() func(m *proto.Message) {
+	return func(m *proto.Message) {
+		p[m.Name] = NewProtoMesssageFromMessage(m)
+	}
+}
+
+func (p ProtoMessageMap) GetFieldNum(messageName, fieldName string) int {
+	_, ok := p[messageName]
+	if !ok {
+		p[messageName] = &ProtoMessage{}
+	}
+	fieldNum := p[messageName].GetFieldNum(fieldName)
+	return fieldNum
+}
+
+func (p ProtoMessageMap) RemoveFieldNum(messageName, fieldName string) {
+	_, ok := p[messageName]
+	if !ok {
+		return
+	}
+	p[messageName].RemoveFieldNum(fieldName)
+	return
+}
+
+type ProtoMessage struct {
+	currMaxNum  int
+	droppedNums []int
+	fields      map[string]int
+}
+
+func NewProtoMesssageFromMessage(msg *proto.Message) *ProtoMessage {
+	fieldsMap := make(map[string]int, len(msg.Elements))
+	currMax := 0
+
+	for i, _ := range msg.Elements {
+		element := msg.Elements[i]
+
+		switch element.(type) {
+		case *proto.NormalField:
```

Comment:
Nitpicky but is normal field just a top level field and map field a nested message field? Normal may be to subjective and general for a name.

Reply:
This is actually imported from elsewhere. But a normal field is basically any kind of field. A map field, I believe, is a specific map<keyType, valueType> in proto syntax

Reply:
Ah cool